imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -1004,3 +1045,247 @@ def create_datasets(
+            h_original = l1b_histrates["h_counts"].values[epoch_idx, az_idx, :]
+            o_original = l1b_histrates["o_counts"].values[epoch_idx, az_idx, :]
+
+            for orig_idx, true_esa_step in energy_step_mapping.items():
+                if orig_idx < len(h_original) and 1 <= true_esa_step <= 7:
+                    reswept_idx = true_esa_step - 1
+                    h_counts_reswept[epoch_idx, az_idx, reswept_idx] += h_original[
+                        orig_idx
+                    ]
+                    o_counts_reswept[epoch_idx, az_idx, reswept_idx] += o_original[
+                        orig_idx
+                    ]
+                    if orig_idx != reswept_idx:
+                        exposure_factor[epoch_idx, az_idx, reswept_idx] += 1
+
+    l1b_histrates["h_counts"].values = h_counts_reswept
+    l1b_histrates["o_counts"].values = o_counts_reswept
+    l1b_histrates.attrs["energy_step_correction"] = (
+        "Applied LUT table energy step mapping"
+    )
+
+    return l1b_histrates, exposure_factor
+
+
+def calculate_histogram_rates(
+    l1b_histrates: xr.Dataset,
+    acq_start: xr.DataArray,
+    acq_end: xr.DataArray,
+    avg_spin_durations_per_cycle: xr.DataArray,
+    exposure_factor: np.ndarray,
+) -> xr.Dataset:
+    """
+    Calculate histogram rates by dividing reswept counts by exposure time.
+
+    For each epoch in l1b_histrates, this function finds the corresponding
+    spin interval, calculates the exposure time for 6-degree bins,
+    and divides the counts by the exposure time. The exposure time is scaled
+    by the number of ESA steps that were reswept during resweeping.
+
+    Parameters
+    ----------
+    l1b_histrates : xr.Dataset
+        The L1B histogram rates dataset containing reswept h_counts and o_counts.
+    acq_start : xr.DataArray
+        Start times for each spin cycle in MET seconds.
+    acq_end : xr.DataArray
+        End times for each spin cycle in MET seconds.
+    avg_spin_durations_per_cycle : xr.DataArray
+        Average spin duration for each cycle in seconds.
+    exposure_factor : np.ndarray
+        3D array of exposure factors (epoch, azimuth, esa_step) indicating how many
+        ESA steps were reswept during resweeping.
+
+    Returns
+    -------
+    l1b_histrates : xr.Dataset
+        Updated dataset with h_rates and o_rates added.
+    """
+    epochs = l1b_histrates["epoch"].values
+    h_counts = l1b_histrates["h_counts"].values
+    o_counts = l1b_histrates["o_counts"].values
+
+    h_rates = np.zeros_like(h_counts, dtype=float)
+    o_rates = np.zeros_like(o_counts, dtype=float)
+    num_azimuth = h_counts.shape[1]
+    exposure_times = np.zeros((len(epochs), num_azimuth, 7), dtype=float)
+
+    for epoch_idx, epoch in enumerate(epochs):
+        spin_cycle_mask = (epoch >= met_to_ttj2000ns(acq_start.values)) & (
+            epoch <= met_to_ttj2000ns(acq_end.values)
+        )
+        spin_cycle_indices = np.where(spin_cycle_mask)[0]
```

Comment:
`np.nonzero()` is preferred when used like this. https://numpy.org/doc/2.3/reference/generated/numpy.where.html
```suggestion
        spin_cycle_indices = np.nonzero(spin_cycle_mask)[0]
```

Reply:
Oh yeah, I think you told me this before and I forgot. Thanks!